fault/random.py

```diff
@@ -1,10 +1,246 @@
+import typing as tp
 import random
+import itertools as it
+from hwtypes import AbstractBitVector, AbstractBit
 from hwtypes import BitVector, Bit
+from hwtypes import z3BitVector, z3Bit
+from collections.abc import Mapping
+import z3
 
+def constrained_random_bv(width, pred):
+    while True:
+        randval = random_bv(width)
+        if pred(randval):
+            return randval
 
 def random_bv(width):
     return BitVector[width](random.randint(0, (1 << width) - 1))
 
-
 def random_bit():
     return Bit(random.randint(0, 1))
+
+class HashableDict(Mapping):
+    def __init__(self, d=None):
+        if d is None:
+            d = dict()
+        self._d = dict(d)
+        self._hash = None
+
+    def __getitem__(self, idx):
+        return self._d[idx]
+
+    def __iter__(self):
+        return iter(self._d)
+
+    def __len__(self):
+        return len(self._d)
+
+    def __hash__(self):
+        if self._hash is None:
+            self._hash = hash(frozenset(self.items()))
+        return self._hash
```

Comment:
It seems like the hash value for an instance is computed and fixed the first time `__hash__` is invoked, is this the intended behavior? What happens if we add items to the dictionary after hash has been called? Or is that a use case that we're not intending to support for this? If so, we should add some docs/guards against this (e.g. calling `__setitem__` after `_hash` has been set could raise an error).

Reply:
Also, `__setitem__` is not defined, but looking at the rest of the code, it seems like this is used to construct a hashable version of a dictionary that isn't modified, so perhaps `FrozenDict` might be a better name?

Reply:
__setitem__ is not defined to make the object immutable.  Changing the value of a hashable object will cause strange behavior as they will not be rehashed when they change.  Will change name to FrozenDict